File: Easy/0001_two-sum/two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
#======== <Solution 1> ========#
        from itertools import combinations
        for (i, j) in combinations(range(len(nums)), 2):
            if nums[i] + nums[j] == target:
                return [i, j]
        
#======== <Solution 2> ========#
        while nums:
            num = nums.pop()
            if target - num in nums:
                return [nums.index(target - num), len(nums)]

#======== <Solution 3> ========#
        seen = {}
        for i, num in enumerate(nums):
            if target - num in seen:
                return [seen[target - num], i]
            else:
                seen[num] = i

File: Easy/1_two-sum/two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
/*======== <Solution 1> ========*/
    while (nums.length > 0) {
        let id = nums.indexOf(target - nums.pop());
        if (id >= 0) {
            return [id, nums.length];
        }
    }
    
/*======== <Solution 2> ========*/
    let memory = {};
    for (let i = 0; i < nums.length; i++) {
        // target = nums[i] + other_num
        // Check if other_num is in memory object
        if (memory[target - nums[i]] != undefined) {
            return [memory[target - nums[i]], i];
        }
        // Store other_num and its index for each round
        memory[nums[i]] = i;
    }
};
